indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,685 @@
+   * with a non-null value or with a non-null value.
+   */
+  public Future<SegmentIdWithShardSpec> add(SegmentAllocateRequest request)
+  {
+    if (!leaderSelector.isLeader()) {
+      throw new ISE("Cannot allocate segment if not leader.");
+    } else if (!isEnabled()) {
+      throw new ISE("Batched segment allocation is disabled.");
+    }
+
+    final AllocateRequestKey requestKey = new AllocateRequestKey(request, false);
+    final AtomicReference<Future<SegmentIdWithShardSpec>> requestFuture = new AtomicReference<>();
+
+    keyToBatch.compute(requestKey, (key, existingBatch) -> {
+      AllocateRequestBatch computedBatch = existingBatch;
+      if (computedBatch == null) {
+        computedBatch = new AllocateRequestBatch(key);
+        computedBatch.resetQueueTime();
+        processingQueue.offer(computedBatch);
+      }
+
+      // Possible race condition:
+      // t1 -> new batch is added to queue or batch already exists in queue
+      // t2 -> executor pops batch, processes all requests in it
+      // t1 -> new request is added to dangling batch and is never picked up
+      // Solution: For existing batch, call keyToBatch.remove() on the key to
+      // wait on keyToBatch.compute() to finish before proceeding with processBatch().
+      // For new batch, keyToBatch.remove() would not wait as key is not in map yet
+      // but a new batch is unlikely to be due immediately, so it won't get popped right away.
+      requestFuture.set(computedBatch.add(request));
+      return computedBatch;
+    });
+
+    return requestFuture.get();
+  }
+
+  private void requeueBatch(AllocateRequestBatch batch)
+  {
+    log.info("Requeueing [%d] failed requests in batch [%s].", batch.size(), batch.key);
+    keyToBatch.compute(batch.key, (key, existingBatch) -> {
+      if (existingBatch == null) {
+        batch.resetQueueTime();
+        processingQueue.offer(batch);
+        return batch;
+      } else {
+        // Merge requests from this batch to existing one
+        existingBatch.merge(batch);
+        return existingBatch;
+      }
+    });
+  }
+
+  private void processBatchesDue()
+  {
+    // If not leader, clear the queue and do not schedule any more rounds of processing
+    if (!leaderSelector.isLeader()) {
+      log.info("Not leader anymore. Clearing [%d] batches from queue.", processingQueue.size());
+      processingQueue.clear();
+      keyToBatch.clear();
+      return;
+    }
+
+    // Process all batches which are due
+    log.debug("Processing all batches which are due for execution.");
+    int numProcessedBatches = 0;
+
+    AllocateRequestBatch nextBatch = processingQueue.peek();
+    while (nextBatch != null && nextBatch.isDue()) {
+      processingQueue.poll();
+      boolean processed;
+      try {
+        processed = processBatch(nextBatch);
+        ++numProcessedBatches;
+      }
+      catch (Throwable t) {
+        processed = true;
```

Comment:
this should be false, right? so that we can requeue it.

Reply:
I am not too sure about this one. An exception being caught here means that there is probably some bug in the code. Retrying wouldn't help. We could maybe raise an alert here? Let me know what you think.

Reply:
well there could be db failures for example. If we don't retry, we should surface the failure to the caller.

Reply:
We can complete the futures exceptionally. That would also be in line with what the current code would do in case of an exception. Thanks for pointing this out.